source/validate.cpp

```diff
@@ -448,22 +450,38 @@ spv_result_t ModuleLayoutPass(ValidationState_t& _,
   if (_.is_enabled(SPV_VALIDATE_LAYOUT_BIT)) {
     SpvOp opcode = inst->opcode;
 
-    if (libspirv::ModuleLayoutSection::kModule == _.getLayoutStage()) {
+    if (_.getLayoutStage() < kLayoutFunctionDeclarations) {
       // Module scoped instructions are processed by determining if the opcode
       // is part of the current stage. If it is not then the next stage is
       // checked.
       while (_.isOpcodeInCurrentLayoutStage(opcode) == false) {
-        // TODO(umar): Check if the MemoryModel instruction has executed
         _.progressToNextLayoutStageOrder();
-        if (_.getLayoutStage() == libspirv::ModuleLayoutSection::kFunction) {
+
+        if (_.getLayoutStage() == kLayoutMemoryModel &&
```

Comment:
Perhaps the error is misleading. This condition takes advantage of the fact that the memory model instruction is required. Therefore if the memory model instruction does not appear at the kLayoutMemoryModel stage then either the instruction does not exist or the layout of instruction is incorrect.